relayer/cli/src/commands/query/channel.rs

```diff
@@ -95,25 +103,27 @@ impl Runnable for QueryChannelEndCmd {
         status_info!("Options", "{:?}", opts);
 
         // run with proof:
-        // cargo run --bin relayer -- -c simple_config.toml query channel end ibc0 transfer ibconexfer
+        // cargo run --bin relayer -- -c relayer/relay/tests/config/fixtures/simple_config.toml query channel end ibc-test firstport firstchannel --height 3
         //
         // run without proof:
-        // cargo run --bin relayer -- -c simple_config.toml query channel end ibc0 transfer ibconexfer -p false
+        // cargo run --bin relayer -- -c relayer/relay/tests/config/fixtures/simple_config.toml query channel end ibc-test firstport firstchannel --height 3 -p false
         //
         // Note: currently both fail in amino_unmarshal_binary_length_prefixed().
         // To test this start a Gaia node and configure a channel using the go relayer.
         let chain = TendermintChain::from_config(chain_config).unwrap();
-        let res = block_on(query_channel(
+        let res = block_on(query::<TendermintChain, ProtoChannel, ChannelEnd>(
             &chain,
-            opts.height,
-            opts.port_id.clone(),
-            opts.channel_id.clone(),
-            opts.proof,
+            Request {
+                path: Some(TendermintPath::from_str(&"store/ibc/key").unwrap()),
+                data: ChannelEndsPath::new(opts.port_id, opts.channel_id)
+                    .to_string()
+                    .into_bytes(),
+                height: Some(Height::from(opts.height)),
+                prove: opts.proof,
+            },
```

Comment:
Can we discuss about this as I am not entirely sure about the change here and in similar places. IMO it would still be nice to have a function like `query_channel` that takes the chain, height, proof and channel key. This function could be called from the CLI or as part of relaying action. The peculiarities of how the query is done depending on the chain type should be in the chain implementation.

Reply:
Good find. :) This part of the code is admittedly not ready yet, it's an example implementation so we can see the whole code running and to be able to highlight the chain-specific instructions. (This was all over the place generating redundant code.)
As a minimum, the Request struct would need a constructor (`::new`). Then, the store path should be stored somewhere so we don't hardwire it in each query. Then we should have some kind of helper function so we don't do this horrible `.to_string().into_bytes()` stuff, which is there because the `Request` struct is just not optimal. The Request struct should come from the tendermint RPC client, which it can't today because it's not fully public there. (Hence it's a copy, which is also a horrible idea.)

But apart from too much wiring shown here, it has exactly the properties you're looking for with one addition: you can customize it to any query. A `query_channel` function's real value is that you don't have to remember the path (`ChannelEndsPath::new(opts.port_id, opts.channel_id)`) but everything else is the same even in another `query_connection` function. This will be specific to every query, so if we write a `query_channel`, we'll have to write it for all queries.

I guess it would be nice to see how this works with a more complex query, like with the client. Client already implements helper functions, maybe implementing it using these tools would surface the optimal ratio for helper functions vs generic functions.

Reply:
To understand why I left this so ugly: I was focusing on the lower part of the `query` function, which was calling `build_response` and it had to be implemented for each query separately. Now it's generic enough that it's only implemented once and you don't have to do it for each query.

Reply:
More notes that I was going to mention on the catchup call: Why this specific implementation?

I wanted to separate the generics of an abci_query from the actual important information, the `data`. An abci_query is always the same, only its input is different each time. The same with the response: we really only care about the data in it. I think the current code is a good enough example of the response part.

The request part is up for discussion. But I would like to keep it slightly generic so we don't produce code that has to be copy-pasted.

Reply:
Here's an alternative approach where the formatting for the request is kept within the QueryXYZOptions struct by using `Into<>`. https://github.com/informalsystems/ibc-rs/compare/greg/query-channel-and-reduce-code...greg/different-query-input

The point is that the request will always require standardized parameters (within a chain), so having different structs for different queries only make sense if we can standardize their content for the RPC call.